api/src/main/java/org/openmrs/module/locationbasedaccess/EncounterSearchAdviser.java

```diff
@@ -0,0 +1,137 @@
+                return null;
+            }
+
+            Object object = invocation.proceed();
+            if (Daemon.isDaemonUser(Context.getAuthenticatedUser()) || Context.getAuthenticatedUser().isSuperUser()) {
+                return object;
+            }
+
+            Integer sessionLocationId = Context.getUserContext().getLocationId();
+            if (sessionLocationId != null) {
+                String sessionLocationUuid = Context.getLocationService().getLocation(sessionLocationId).getUuid();
+
+                if (object instanceof List) {
+                    List<Encounter> encounterList = (List<Encounter>) object;
+                    object = removeEncountersIfNotBelongToGivenLocation(encounterList, sessionLocationUuid);
+                } else if (object instanceof Map) {
+                    Map<Integer, List<Encounter>> encounterMap = (Map<Integer, List<Encounter>>) object;
+                    Iterator<Map.Entry<Integer, List<Encounter>>> mapIterator = encounterMap.entrySet().iterator();
+                    while (mapIterator.hasNext()) {
+                        Map.Entry<Integer, List<Encounter>> entry = mapIterator.next();
+                        List<Encounter> encounterList = entry.getValue();
+                        entry.setValue(removeEncountersIfNotBelongToGivenLocation(encounterList, sessionLocationUuid));
+                        //TODO: remove the entry from the map, if the encounter list is empty and update the map index
+                    }
+                    object = encounterMap;
+                } else if (object instanceof Encounter) {
+                    if (!doesEncounterBelongToGivenLocation((Encounter) object, sessionLocationUuid)) {
+                        object = null;
+                    }
+                }
+            } else {
+                log.debug("Search Encounter : Null Session Location in the UserContext");
+                if (object instanceof Encounter) {
+                    // If the sessionLocationId is null, then return null for a Encounter instance
+                    return null;
+                } else if (object instanceof Map) {
+                    // If the sessionLocationId is null, then return a empty map
+                    return new HashMap<Integer, List<Encounter>>();
+                } else {
+                    // If the sessionLocationId is null, then return a empty list
+                    return new ArrayList<Encounter>();
+                }
+            }
+            return object;
+        }
+    }
+
+    private List<Encounter> removeEncountersIfNotBelongToGivenLocation(List<Encounter> encounterList, String sessionLocationUuid) {
+        for (Iterator<Encounter> iterator = encounterList.iterator(); iterator.hasNext(); ) {
+            Encounter thisEncounter = iterator.next();
+            if(!doesEncounterBelongToGivenLocation(thisEncounter, sessionLocationUuid)) {
+                iterator.remove();
```

Comment:
Doesn't this remove land you in problems of manipulating a collection in the process of iterating over it?

Reply:
Nope, this is a safety method to remove from the collections, and I used this in the existing implementations as well.